runtime/common/src/impl_on_charge_evm_transaction.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#[macro_export]
+macro_rules! impl_on_charge_evm_transaction {
+	{} => {
+		type CurrencyAccountId<T> = <T as frame_system::Config>::AccountId;
+
+		type BalanceFor<T> =
+			<<T as pallet_evm::Config>::Currency as CurrencyT<CurrencyAccountId<T>>>::Balance;
+
+		type PositiveImbalanceFor<T> =
+			<<T as pallet_evm::Config>::Currency as CurrencyT<CurrencyAccountId<T>>>::PositiveImbalance;
+
+		type NegativeImbalanceFor<T> =
+			<<T as pallet_evm::Config>::Currency as CurrencyT<CurrencyAccountId<T>>>::NegativeImbalance;
+
+		pub struct OnChargeEVMTransaction<OU>(sp_std::marker::PhantomData<OU>);
+		impl<T, OU> OnChargeEVMTransactionT<T> for OnChargeEVMTransaction<OU>
+		where
+			T: pallet_evm::Config,
+			PositiveImbalanceFor<T>: Imbalance<BalanceFor<T>, Opposite = NegativeImbalanceFor<T>>,
+			NegativeImbalanceFor<T>: Imbalance<BalanceFor<T>, Opposite = PositiveImbalanceFor<T>>,
+			OU: OnUnbalanced<NegativeImbalanceFor<T>>,
+		{
+			type LiquidityInfo = Option<NegativeImbalanceFor<T>>;
+
+			fn withdraw_fee(who: &H160, fee: U256) -> Result<Self::LiquidityInfo, pallet_evm::Error<T>> {
+				EVMCurrencyAdapter::<<T as pallet_evm::Config>::Currency, ()>::withdraw_fee(who, fee)
+			}
+
+			fn correct_and_deposit_fee(
+				who: &H160,
+				corrected_fee: U256,
+				already_withdrawn: Self::LiquidityInfo,
+			) {
+				<EVMCurrencyAdapter<<T as pallet_evm::Config>::Currency, OU> as OnChargeEVMTransactionT<
+					T,
+				>>::correct_and_deposit_fee(who, corrected_fee, already_withdrawn)
+			}
+
+			fn pay_priority_fee(_tip: U256) {}
```

Comment:
It amuses me how subtle this is in code...

Reply:
Is it necessary to add some docs about this?

Reply:
@AsceticBear PTAL to https://github.com/paritytech/frontier/pull/700

I thought about it this past days, and decided to slightly modify the approach in Frontier. I think splitting both base fee and priority fee handling offers the most flexibility in terms of how each chain wants to deals with them.